src/payload.rs

```diff
@@ -776,25 +904,51 @@ impl PayloadSnapshot {
     }
 
     /// Returns an slice of the payload.
-    pub fn payload(&self) -> &[(Payload, OriginInfo)] {
+    pub fn payload(&self) -> &[(Payload, PayloadInfo)] {
         &self.payload
     }
 
     /// Returns an iterator over the route origins.
     pub fn origins(
         &self
-    ) -> impl Iterator<Item = (RouteOrigin, &OriginInfo)> + '_ {
-        self.payload().iter().filter_map(|item| {
+    ) -> impl Iterator<Item = (RouteOrigin, &PayloadInfo)> + '_ {
+        self.payload()[..self.router_key_index].iter().filter_map(|item| {
```

Comment:
The `filter_` part seems unnecessary assuming all `item`s up until index `self.router_key_index` are of the `Payload::Origin` enum variant anyway. Could a `map` suffice, e.g.
```rust
.iter().map(|item| {
    (item.0.origin, &item.1)
}
```
If not mistaking, the final `None` should come from `iter()` anyway.

Reply:
This has now all been rewritten. See comment below.